### DISCUSSION TITLE
Wrap lines around defined edge

### DISCUSSION THREAD

**Discussion by smartmic** [orbitalquark/textadept#37] [Q&A]
The `Ctrl-Alt-\` key binding visually wraps long lines around the window border. But how can I hard wrap (with newlines) paragraphs around a fixed column position like 72 or set this as standard editing option? I saw many options around wrapping in the API docs but could not figure out a way to do so with those…

**Comment by orbitalquark** [comment]
Unfortunately, Scintilla[1], the editing component Textadept uses, does not allow arbitrary column wrapping. The best you can do is resize the window to 72 columns or resize a split view to 72 columns.

[1]: https://scintilla.org

**Reply by smartmic** [reply]
Alright, got it! 

Interesting though that Scintilla lacks this functionality, I thought this is a quite common feature, at least for terminal editors.

**Comment by rgieseke** [comment]
For hard-wrapping with an actual newline you could write a function that gets the current line (or selection or paragraph) and then replaces it with the reindented text.
I used to have a function like this, but couldn't find it again, nor in the mailing list archive or wiki.

If you want this for continuous writing you could hook into the `char_added` event, and add a newline after a space if the current column position is >72.

**Comment by rgieseke** [comment]
For the latter something like
```lua
events.connect(events.CHAR_ADDED, function(code)
  if code == string.byte(" ") then
    if buffer.column[buffer.current_pos] > 72 then
      buffer.add_text('\n')
    end
  end
end)
```
But you probably want to wrap at any char and go back to the last space and insert the newline there.

**Reply by smartmic** [reply]
Good idea, thanks! I will try this later and hope that this function does not add any notable latency when typing fast.

**Reply by rgieseke** [reply]
If you want to cut off a few bit flips you could also compare directly to the code given instead of the conversion ...

**Comment by orbitalquark** [comment]
Scintilla was written to be a GUI editing component. I was the one that wrote a terminal interface to it[1]. Since it was never designed for the terminal, there will likely be other quirks like this.

[1]: https://orbitalquark.github.io/scinterm

**Reply by smartmic** [reply]
Never mind the quirks, having a terminal version of textadept is a BIG plus for me. So thank you for all the effort you put into scinterm and textadept, it is really great!

**Comment by smartmic** [comment]
@rgieseke your solution works fine, no lagging etc. 

But I solved the formatting using an external perl filter, so I do not have to fiddle with a proper wrapping algorithm. The package is [Text::Autoformat](https://metacpan.org/pod/Text::Autoformat) and provides some more features for formatting non-trivial paragraphs. 

So my current setup is using a split view resized to ~72 columns as suggested by @orbitalquark and, if necessary, doing a hard wrap reformatting using this custom key binding:

```lua
-- format paragraph
keys['alt+p'] = function() 
	textadept.editing.select_paragraph()
	textadept.editing.filter_through("perl -MText::Autoformat -e'autoformat'")
end
```

It is really cool to extend textadept with such few lines of configuration!

**Reply by rgieseke** [reply]
Great solution, will try that, too!

**Reply by goober99** [reply]
I did something similar but using the fmt command instead of Perl Text::Autoformat:

```
-- Reflow lines bound to ctrl+j like Geany
keys['ctrl+j'] = function()
  if buffer:get_sel_text() == '' then textadept.editing.select_paragraph() end
  local edge = edge_column or 80
  textadept.editing.filter_through('fmt -w' .. edge .. ' -g' .. edge)
end
```
